taxon_filter.py

```diff
@@ -554,9 +529,89 @@ def wrapper(inBam, db, outBam, threads, JVMmemory=None):
     return 0
+__commands__.append(('deplete_blastn_bam', parser_deplete_blastn_bam))
 
 
-__commands__.append(('deplete_blastn_bam', parser_deplete_blastn_bam))
+@contextlib.contextmanager
+def extract_build_or_use_database(db, db_build_command, db_extension_to_expect, tmp_suffix='db_unpack', db_prefix="db"):
+    '''
+    db_extension_to_expect = file extension, sans dot prefix
+    '''
+    with util.file.tmp_dir(tmp_suffix) as tempDbDir:
+        db_dir = ""
+        if os.path.exists(db):
+            if os.path.isfile(db):
+                # this is a single file
+                if db.endswith('.fasta') or db.endswith('.fasta.gz') or db.endswith('.fasta.lz4') or db.endswith('.fa') or db.endswith('.fa.gz') or db.endswith('.fa.lz4'):
+                    # this is an unindexed fasta file, we will need to index it
+                    # function should conform to the signature: 
+                    # db_build_command(inputFasta, outputDirectory, outputFilePrefix)
+                    # the function will need to be able to handle lz4, etc.
+                    db_build_command(db, tempDbDir, db_prefix)
+                    db_dir = tempDbDir
+                else:
+                    # this is a tarball with prebuilt indexes
+                    db_dir = util.file.extract_tarball(db, tempDbDir)
+            else:
+                # this is a directory
+                db_dir = db
+            # this directory should have a .nin file
+            hits = list(glob.glob(os.path.join(db_dir, '*.{ext}'.format(ext=db_extension_to_expect))))
+            if len(hits) == 0:
+                raise Exception("The blast database does not appear to a *.{ext} file.".format(ext=db_extension_to_expect))
+            elif len(hits) == 1:
+                db_prefix = hits[0][:-(len('.{ext}'.format(ext=db_extension_to_expect)))]  # remove the '.extension'
+            elif len(hits) >1:
+                db_prefix = os.path.commonprefix(hits).rsplit('.', 1)[0] # remove extension and split-db prefix
+        else:
+            # this is simply a prefix to a bunch of files, not an actual file
+            db_prefix = db
+
+        yield (db_prefix,tempDbDir)
+
+# ========================
+# ***  deplete_bwa  ***
+# ========================
+
+def deplete_bwa_bam(inBam, db, outBam, threads=None, JVMmemory=None):
+    'Use blastn to remove reads that match at least one of the databases.'
+
+    threads = util.misc.sanitize_thread_count(threads)
+
+    with extract_build_or_use_database(db, bwa_build_db, 'bwt', tmp_suffix="-bwa_db_unpack", db_prefix="bwa") as (db_prefix,tempDbDir):
+        with util.file.tempfname('.aligned.sam') as aligned_sam:
+            tools.bwa.Bwa().align_mem_bam(inBam, db_prefix, aligned_sam, threads=threads)
+            tools.samtools.SamtoolsTool().view(['-f0x4'], aligned_sam, outBam)
+            # TODO: consider using Bwa().mem() so the input bam is not broken out by read group
```

Comment:
All good TODO thoughts. As for the first one, the question is whether `bwa mem` mangles the RG headers (I think it does.. I think that's why we have the custom function. Since `deplete_bwa_bam` actually emits output that was, at some point, funneled through BWA itself (vs all the other depletion methods that just take inBam->outBam via Picard FilterSamReads), it's actually important to make sure that BWA mem doesn't mangle the headers.